--- a/net/minecraft/server/BlockCauldron.java
+++ b/net/minecraft/server/BlockCauldron.java
@@ -1,5 +1,7 @@
 package net.minecraft.server;
 
+import org.bukkit.event.block.CauldronLevelChangeEvent; // CraftBukkit
+
 public class BlockCauldron extends BlockCauldronAbstract {
 
     public BlockCauldron(BlockBase.Info blockbase_info) {
@@ -38,4 +40,31 @@
         }
 
     }
+
+    // CraftBukkit start
+    static boolean changeLevel(World world, BlockPosition blockposition, IBlockData iblockdata, int i, Entity entity, CauldronLevelChangeEvent.ChangeReason reason) {
+        int newLevel = Integer.valueOf(MathHelper.clamp(i, 0, 3));
+        CauldronLevelChangeEvent event = new CauldronLevelChangeEvent(
+                world.getWorld().getBlockAt(blockposition.getX(), blockposition.getY(), blockposition.getZ()),
+                (entity == null) ? null : entity.getBukkitEntity(), reason, iblockdata.get(LayeredCauldronBlock.b), newLevel
+        );
+        world.getServer().getPluginManager().callEvent(event);
+        if (event.isCancelled()) {
+            return false;
+        }
+        world.setTypeAndData(blockposition, (IBlockData) iblockdata.set(LayeredCauldronBlock.b, event.getNewLevel()), 2);
+        world.updateAdjacentComparators(blockposition, iblockdata.getBlock());
+        return true;
+    }
+
+    static boolean changeLevelIncrease(World world, BlockPosition blockposition, IBlockData iblockdata, Entity entity, CauldronLevelChangeEvent.ChangeReason reason) {
+        int i = (Integer) iblockdata.get(LayeredCauldronBlock.b) + 1;
+        return changeLevel(world, blockposition, iblockdata, i, entity, reason);
+    }
+
+    static boolean changeLevelDecrease(World world, BlockPosition blockposition, IBlockData iblockdata, Entity entity, CauldronLevelChangeEvent.ChangeReason reason) {
+        int i = (Integer) iblockdata.get(LayeredCauldronBlock.b) - 1;
+        return changeLevel(world, blockposition, iblockdata, i, entity, reason);
+    }
+    // CraftBukkit end
 }
