--- a/net/minecraft/server/ChunkRegionLoader.java
+++ b/net/minecraft/server/ChunkRegionLoader.java
@@ -92,7 +92,7 @@
 
             if (nbttagcompound1.hasKeyOfType("TileTicks", 9)) {
                 nbttaglist1 = nbttagcompound1.getList("TileTicks", 10);
-                function = IRegistry.BLOCK::getKey;
+                function = (block) -> IRegistry.BLOCK.getKey((Block) block); // CraftBukkit - decompile error
                 registryblocks = IRegistry.BLOCK;
                 registryblocks.getClass();
                 object1 = TickListChunk.a(nbttaglist1, function, registryblocks::get);
@@ -104,7 +104,7 @@
 
             if (nbttagcompound1.hasKeyOfType("LiquidTicks", 9)) {
                 nbttaglist1 = nbttagcompound1.getList("LiquidTicks", 10);
-                function = IRegistry.FLUID::getKey;
+                function = (fluid) -> IRegistry.FLUID.getKey((FluidType) fluid); // CraftBukkit - decompile error
                 registryblocks = IRegistry.FLUID;
                 registryblocks.getClass();
                 object2 = TickListChunk.a(nbttaglist1, function, registryblocks::get);
@@ -114,6 +114,12 @@
 
             object = new Chunk(worldserver.getMinecraftWorld(), chunkcoordintpair, biomestorage, chunkconverter, (TickList) object1, (TickList) object2, k, achunksection, (chunk) -> {
                 loadEntities(worldserver, nbttagcompound1, chunk);
+                // CraftBukkit start - load chunk persistent data from nbt
+                NBTBase persistentBase = nbttagcompound1.get("ChunkBukkitValues");
+                if (persistentBase instanceof NBTTagCompound) {
+                    chunk.persistentDataContainer.putAll((NBTTagCompound) persistentBase);
+                }
+                // CraftBukkit end
             });
         } else {
             ProtoChunk protochunk = new ProtoChunk(chunkcoordintpair, chunkconverter, achunksection, protochunkticklist, protochunkticklist1, worldserver);
@@ -178,7 +184,13 @@
         }
 
         if (chunkstatus_type == ChunkStatus.Type.LEVELCHUNK) {
-            return new ProtoChunkExtension((Chunk) object);
+            // CraftBukkit start - store chunk persistent data in nbt
+            Chunk chunk = (Chunk) object;
+            if (!chunk.persistentDataContainer.isEmpty()) {
+                nbttagcompound1.set("ChunkBukkitValues", chunk.persistentDataContainer.toTagCompound());
+            }
+            return new ProtoChunkExtension(chunk);
+            // CraftBukkit end
         } else {
             ProtoChunk protochunk1 = (ProtoChunk) object;
 
@@ -245,8 +257,9 @@
         boolean flag = ichunkaccess.t();
 
         for (int i = lightenginethreaded.c(); i < lightenginethreaded.d(); ++i) {
+            int finalI = i; // CraftBukkit - decompile errors
             ChunkSection chunksection = (ChunkSection) Arrays.stream(achunksection).filter((chunksection1) -> {
-                return chunksection1 != null && SectionPosition.a(chunksection1.getYPosition()) == i;
+                return chunksection1 != null && SectionPosition.a(chunksection1.getYPosition()) == finalI; // CraftBukkit - decompile errors
             }).findFirst().orElse(Chunk.a);
             NibbleArray nibblearray = lightenginethreaded.a(EnumSkyBlock.BLOCK).a(SectionPosition.a(chunkcoordintpair, i));
             NibbleArray nibblearray1 = lightenginethreaded.a(EnumSkyBlock.SKY).a(SectionPosition.a(chunkcoordintpair, i));
