--- a/net/minecraft/server/EntityArmorStand.java
+++ b/net/minecraft/server/EntityArmorStand.java
@@ -5,6 +5,15 @@
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
 
+// CraftBukkit start
+import org.bukkit.inventory.EquipmentSlot;
+import org.bukkit.craftbukkit.CraftEquipmentSlot;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.entity.ArmorStand;
+import org.bukkit.entity.Player;
+import org.bukkit.event.player.PlayerArmorStandManipulateEvent;
+// CraftBukkit end
+
 public class EntityArmorStand extends EntityLiving {
 
     private static final Vector3f be = new Vector3f(0.0F, 0.0F, 0.0F);
@@ -55,6 +64,13 @@
         this.setPosition(d0, d1, d2);
     }
 
+    // CraftBukkit start - SPIGOT-3607, SPIGOT-3637
+    @Override
+    public float getBukkitYaw() {
+        return this.yaw;
+    }
+    // CraftBukkit end
+
     @Override
     public void updateSize() {
         double d0 = this.locX();
@@ -353,6 +369,21 @@
             return false;
         } else {
             ItemStack itemstack2;
+            // CraftBukkit start
+            org.bukkit.inventory.ItemStack armorStandItem = CraftItemStack.asCraftMirror(itemstack1);
+            org.bukkit.inventory.ItemStack playerHeldItem = CraftItemStack.asCraftMirror(itemstack);
+
+            Player player = (Player) entityhuman.getBukkitEntity();
+            ArmorStand self = (ArmorStand) this.getBukkitEntity();
+
+            EquipmentSlot slot = CraftEquipmentSlot.getSlot(enumitemslot);
+            PlayerArmorStandManipulateEvent armorStandManipulateEvent = new PlayerArmorStandManipulateEvent(player,self,playerHeldItem,armorStandItem,slot);
+            this.world.getServer().getPluginManager().callEvent(armorStandManipulateEvent);
+
+            if (armorStandManipulateEvent.isCancelled()) {
+                return true;
+            }
+            // CraftBukkit end
 
             if (entityhuman.eU().canInstantlyBuild && itemstack1.isEmpty() && !itemstack.isEmpty()) {
                 itemstack2 = itemstack.cloneItemStack();
@@ -381,9 +412,19 @@
     public boolean damageEntity(DamageSource damagesource, float f) {
         if (!this.world.isClientSide && !this.cU()) {
             if (DamageSource.OUT_OF_WORLD.equals(damagesource)) {
-                this.killEntity();
+                // CraftBukkit start
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.handleNonLivingEntityDamageEvent(this, damagesource, f)) {
+                    return false;
+                }
+                // CraftBukkit end
+                this.killEntity(); // CraftBukkit - this.die() -> this.killEntity()
                 return false;
-            } else if (!this.isInvulnerable(damagesource) && !this.armorStandInvisible && !this.isMarker()) {
+            } else if (!this.isInvulnerable(damagesource) && (true || !this.armorStandInvisible) && !this.isMarker()) { // CraftBukkit
+                // CraftBukkit start
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.handleNonLivingEntityDamageEvent(this, damagesource, f, true, this.armorStandInvisible)) {
+                    return false;
+                }
+                // CraftBukkit end
                 if (damagesource.isExplosion()) {
                     this.g(damagesource);
                     this.killEntity();
@@ -457,13 +498,13 @@
     }
 
     private void f(DamageSource damagesource) {
-        Block.a(this.world, this.getChunkCoordinates(), new ItemStack(Items.ARMOR_STAND));
+        drops.add(org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(new ItemStack(Items.ARMOR_STAND))); // CraftBukkit - add to drops
         this.g(damagesource);
     }
 
     private void g(DamageSource damagesource) {
         this.F();
-        this.d(damagesource);
+        // this.d(damagesource); // CraftBukkit - moved down
 
         ItemStack itemstack;
         int i;
@@ -471,7 +512,7 @@
         for (i = 0; i < this.handItems.size(); ++i) {
             itemstack = (ItemStack) this.handItems.get(i);
             if (!itemstack.isEmpty()) {
-                Block.a(this.world, this.getChunkCoordinates().up(), itemstack);
+                drops.add(org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(itemstack)); // CraftBukkit - add to drops
                 this.handItems.set(i, ItemStack.b);
             }
         }
@@ -479,10 +520,11 @@
         for (i = 0; i < this.armorItems.size(); ++i) {
             itemstack = (ItemStack) this.armorItems.get(i);
             if (!itemstack.isEmpty()) {
-                Block.a(this.world, this.getChunkCoordinates().up(), itemstack);
+                drops.add(org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(itemstack)); // CraftBukkit - add to drops
                 this.armorItems.set(i, ItemStack.b);
             }
         }
+        this.d(damagesource); // CraftBukkit - moved from above
 
     }
 
@@ -741,4 +783,11 @@
     private EntitySize s(boolean flag) {
         return flag ? EntityArmorStand.bk : (this.isBaby() ? EntityArmorStand.bl : this.getEntityType().m());
     }
+
+    // CraftBukkit start
+    @Override
+    protected boolean isDropExperience() {
+        return true; // MC-157395, SPIGOT-5193 even baby (small) armor stands should drop
+    }
+    // CraftBukkit end
 }
