--- a/net/minecraft/server/EntityLeash.java
+++ b/net/minecraft/server/EntityLeash.java
@@ -4,6 +4,8 @@
 import java.util.List;
 import javax.annotation.Nullable;
 
+import org.bukkit.craftbukkit.event.CraftEventFactory; // CraftBukkit
+
 public class EntityLeash extends EntityHanging {
 
     public EntityLeash(EntityTypes<? extends EntityLeash> entitytypes, World world) {
@@ -68,22 +70,42 @@
             while (iterator.hasNext()) {
                 entityinsentient = (EntityInsentient) iterator.next();
                 if (entityinsentient.getLeashHolder() == entityhuman) {
+                    // CraftBukkit start
+                    if (CraftEventFactory.callPlayerLeashEntityEvent(entityinsentient, this, entityhuman).isCancelled()) {
+                        ((EntityPlayer) entityhuman).playerConnection.sendPacket(new PacketPlayOutAttachEntity(entityinsentient, entityinsentient.getLeashHolder()));
+                        continue;
+                    }
+                    // CraftBukkit end
                     entityinsentient.setLeashHolder(this, true);
                     flag = true;
                 }
             }
 
             if (!flag) {
-                this.ab();
-                if (entityhuman.eX().canInstantlyBuild) {
+                // CraftBukkit start - Move below
+                // this.die();
+                boolean die = true;
+                // CraftBukkit end
+                if (true || entityhuman.eX().canInstantlyBuild) { // CraftBukkit - Process for non-creative as well
                     iterator = list.iterator();
 
                     while (iterator.hasNext()) {
                         entityinsentient = (EntityInsentient) iterator.next();
                         if (entityinsentient.isLeashed() && entityinsentient.getLeashHolder() == this) {
-                            entityinsentient.unleash(true, false);
+                            // CraftBukkit start
+                            if (CraftEventFactory.callPlayerUnleashEntityEvent(entityinsentient, entityhuman).isCancelled()) {
+                                die = false;
+                                continue;
+                            }
+                            entityinsentient.unleash(true, !entityhuman.eX().canInstantlyBuild); // false -> survival mode boolean
+                            // CraftBukkit end
                         }
                     }
+                    // CraftBukkit start
+                    if (die) {
+                        this.killEntity();
+                    }
+                    // CraftBukkit end
                 }
             }
 
