--- a/net/minecraft/server/EntityBee.java
+++ b/net/minecraft/server/EntityBee.java
@@ -138,7 +138,7 @@
                 }
 
                 if (b0 > 0) {
-                    ((EntityLiving) entity).addEffect(new MobEffect(MobEffects.POISON, b0 * 20, 0));
+                    ((EntityLiving) entity).addEffect(new MobEffect(MobEffects.POISON, b0 * 20, 0), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); // CraftBukkit
                 }
             }
 
@@ -511,11 +511,15 @@
         if (this.isInvulnerable(damagesource)) {
             return false;
         } else {
-            if (!this.world.isClientSide) {
+            // CraftBukkit start
+            boolean result = super.damageEntity(damagesource, f);
+
+            if (result && !this.world.isClientSide) {
                 this.bx.l();
             }
 
-            return super.damageEntity(damagesource, f);
+            return result;
+            // CraftBukkit end
         }
     }
 
@@ -536,7 +540,7 @@
     class d extends EntityBee.a {
 
         private d() {
-            super(null);
+            super(); // CraftBukkit - decompile error
         }
 
         @Override
@@ -596,7 +600,7 @@
     class g extends EntityBee.a {
 
         private g() {
-            super(null);
+            super(); // CraftBukkit - decompile error
         }
 
         @Override
@@ -645,7 +649,7 @@
                             }
                         }
 
-                        if (flag) {
+                        if (flag && !org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(EntityBee.this, blockposition, iblockdata.set(blockstateinteger, (Integer) iblockdata.get(blockstateinteger) + 1)).isCancelled()) { // Spigot
                             EntityBee.this.world.triggerEffect(2005, blockposition, 0);
                             EntityBee.this.world.setTypeUpdate(blockposition, (IBlockData) iblockdata.set(blockstateinteger, (Integer) iblockdata.get(blockstateinteger) + 1));
                             EntityBee.this.fD();
@@ -660,7 +664,7 @@
     class i extends EntityBee.a {
 
         private i() {
-            super(null);
+            super(); // CraftBukkit - decompile error
         }
 
         @Override
@@ -724,7 +728,7 @@
         private int h;
 
         k() {
-            super(null);
+            super(); // CraftBukkit - decompile error
             this.a(EnumSet.of(PathfinderGoal.Type.MOVE));
         }
 
@@ -907,7 +911,7 @@
         private int c;
 
         f() {
-            super(null);
+            super(); // CraftBukkit - decompile error
             this.c = EntityBee.this.world.random.nextInt(10);
             this.a(EnumSet.of(PathfinderGoal.Type.MOVE));
         }
@@ -965,7 +969,7 @@
         private int f;
 
         e() {
-            super(null);
+            super(); // CraftBukkit - decompile error
             this.c = EntityBee.this.world.random.nextInt(10);
             this.d = Lists.newArrayList();
             this.a(EnumSet.of(PathfinderGoal.Type.MOVE));
@@ -1184,7 +1188,7 @@
         @Override
         protected void a(EntityInsentient entityinsentient, EntityLiving entityliving) {
             if (entityinsentient instanceof EntityBee && this.e.hasLineOfSight(entityliving)) {
-                entityinsentient.setGoalTarget(entityliving);
+                entityinsentient.setGoalTarget(entityliving, org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_ENTITY, true); // CraftBukkit - reason
             }
 
         }
