--- a/net/minecraft/server/InventoryEnderChest.java
+++ b/net/minecraft/server/InventoryEnderChest.java
@@ -1,14 +1,29 @@
 package net.minecraft.server;
 
 import javax.annotation.Nullable;
+import org.bukkit.Location;
+import org.bukkit.inventory.InventoryHolder;
 
 public class InventoryEnderChest extends InventorySubcontainer {
 
     @Nullable
     private TileEntityEnderChest a;
+    // CraftBukkit start
+    private final EntityHuman owner;
 
-    public InventoryEnderChest() {
+    public InventoryHolder getBukkitOwner() {
+        return owner.getBukkitEntity();
+    }
+
+    @Override
+    public Location getLocation() {
+        return this.a != null ? new Location(this.a.getWorld().getWorld(), this.a.getPosition().getX(), this.a.getPosition().getY(), this.a.getPosition().getZ()) : null;
+    }
+
+    public InventoryEnderChest(EntityHuman owner) {
         super(27);
+        this.owner = owner;
+        // CraftBukkit end
     }
 
     public void a(TileEntityEnderChest tileentityenderchest) {
