--- a/net/minecraft/server/ItemFlintAndSteel.java
+++ b/net/minecraft/server/ItemFlintAndSteel.java
@@ -13,10 +13,18 @@
         BlockPosition blockposition = itemactioncontext.getClickPosition();
         IBlockData iblockdata = world.getType(blockposition);
 
-        if (!BlockCampfire.h(iblockdata) && !BlockCandle.g(iblockdata) && !BlockCandleCake.g(iblockdata)) {
+        if (!BlockCampfire.h(iblockdata) && !BlockCandle.g(iblockdata) && !BlockCandleCake.g(iblockdata) && !org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(world, blockposition, entityhuman).isCancelled()) { // CraftBukkit
             BlockPosition blockposition1 = blockposition.shift(itemactioncontext.getClickedFace());
 
             if (BlockFireAbstract.a(world, blockposition1, itemactioncontext.g())) {
+                // CraftBukkit start - Store the clicked block
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(world, blockposition1, org.bukkit.event.block.BlockIgniteEvent.IgniteCause.FLINT_AND_STEEL, entityhuman).isCancelled()) {
+                    itemactioncontext.getItemStack().damage(1, entityhuman, (entityhuman1) -> {
+                        entityhuman1.broadcastItemBreak(itemactioncontext.getHand());
+                    });
+                    return EnumInteractionResult.PASS;
+                }
+                // CraftBukkit end
                 world.playSound(entityhuman, blockposition1, SoundEffects.FLINTANDSTEEL_USE, SoundCategory.BLOCKS, 1.0F, world.getRandom().nextFloat() * 0.4F + 0.8F);
                 world.a((Entity) entityhuman, ckz.F, blockposition);
                 IBlockData iblockdata1 = BlockFireAbstract.a((IBlockAccess) world, blockposition1);
