--- a/net/minecraft/server/OpList.java
+++ b/net/minecraft/server/OpList.java
@@ -3,7 +3,7 @@
 import com.google.gson.JsonObject;
 import com.mojang.authlib.GameProfile;
 import java.io.File;
-import java.util.Objects;
+import java.util.Iterator;
 
 public class OpList extends JsonList<GameProfile, OpListEntry> {
 
@@ -18,7 +18,7 @@
 
     @Override
     public String[] getEntries() {
-        return (String[]) this.d().stream().map(JsonListEntry::getKey).filter(Objects::nonNull).map(GameProfile::getName).toArray((i) -> {
+        return (String[]) this.d().stream().map(JsonListEntry::getKey).filter(java.util.Objects::nonNull).map(GameProfile::getName).toArray((i) -> {
             return new String[i];
         });
     }
