--- a/net/minecraft/server/RegionLimitedWorldAccess.java
+++ b/net/minecraft/server/RegionLimitedWorldAccess.java
@@ -238,6 +238,13 @@
 
     @Override
     public boolean addEntity(Entity entity) {
+        // CraftBukkit start
+        return addEntity(entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT);
+    }
+
+    @Override
+    public boolean addEntity(Entity entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
+        // CraftBukkit end
         int i = SectionPosition.a(entity.cL());
         int j = SectionPosition.a(entity.cR());
 
