--- a/net/minecraft/server/SoundEffectType.java
+++ b/net/minecraft/server/SoundEffectType.java
@@ -63,10 +63,10 @@
     public static final SoundEffectType ag = new SoundEffectType(1.0F, 1.0F, SoundEffects.COPPER_BREAK, SoundEffects.COPPER_STEP, SoundEffects.COPPER_PLACE, SoundEffects.COPPER_HIT, SoundEffects.COPPER_FALL);
     public final float ah;
     public final float ai;
-    private final SoundEffect aj;
+    public final SoundEffect aj; // PAIL private -> public, rename breakSound
     private final SoundEffect ak;
     private final SoundEffect al;
-    private final SoundEffect am;
+    public final SoundEffect am; // PAIL private -> public, rename hitSound
     private final SoundEffect an;
 
     public SoundEffectType(float f, float f1, SoundEffect soundeffect, SoundEffect soundeffect1, SoundEffect soundeffect2, SoundEffect soundeffect3, SoundEffect soundeffect4) {
