--- a/net/minecraft/server/TileEntityConduit.java
+++ b/net/minecraft/server/TileEntityConduit.java
@@ -6,6 +6,10 @@
 import java.util.Random;
 import java.util.UUID;
 import javax.annotation.Nullable;
+// CraftBukkit start
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+// CraftBukkit end
 
 public class TileEntityConduit extends TileEntity {
 
@@ -177,7 +181,7 @@
                 EntityHuman entityhuman = (EntityHuman) iterator.next();
 
                 if (blockposition.a((BaseBlockPosition) entityhuman.getChunkCoordinates(), (double) j) && entityhuman.isInWaterOrRain()) {
-                    entityhuman.addEffect(new MobEffect(MobEffects.CONDUIT_POWER, 260, 0, true, true));
+                    entityhuman.addEffect(new MobEffect(MobEffects.CONDUIT_POWER, 260, 0, true, true), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.CONDUIT); // CraftBukkit
                 }
             }
 
@@ -206,8 +210,13 @@
         }
 
         if (tileentityconduit.target != null) {
-            world.playSound((EntityHuman) null, tileentityconduit.target.locX(), tileentityconduit.target.locY(), tileentityconduit.target.locZ(), SoundEffects.CONDUIT_ATTACK_TARGET, SoundCategory.BLOCKS, 1.0F, 1.0F);
-            tileentityconduit.target.damageEntity(DamageSource.MAGIC, 4.0F);
+            // CraftBukkit start
+            CraftEventFactory.blockDamage = CraftBlock.at(tileentityconduit.world, tileentityconduit.position);
+            if (tileentityconduit.target.damageEntity(DamageSource.MAGIC, 4.0F)) {
+                tileentityconduit.world.playSound((EntityHuman) null, tileentityconduit.target.locX(), tileentityconduit.target.locY(), tileentityconduit.target.locZ(), SoundEffects.CONDUIT_ATTACK_TARGET, SoundCategory.BLOCKS, 1.0F, 1.0F);
+            }
+            CraftEventFactory.blockDamage = null;
+            // CraftBukkit end
         }
 
         if (entityliving != tileentityconduit.target) {
